app/controllers/api_docs/services_controller.rb

```diff
@@ -41,7 +43,7 @@ def initialize(name, system_name)
 
     def json
       parsed_content = JSON.parse(file_content)
-      parsed_content['basePath'] = backend_base_host if backend_api?
+      parsed_content['servers'] = [{ 'url' => backend_api? ? backend_base_host : '/' }]
```

Comment:
Yes, this is quite ugly, but I don't think we can't do what you suggest...

1. The variable `request` that is used for calculating `base_url` is not available at this point (it is available from the controller actions, but ApiFile is initialized in another place
2. The files are read from file system one time, but then can be accessed from different tenants, so we need to change the base URL for the tenant that it was requested from.
In the original implementation I think rather than the full URL just the path was used.
I needed to set the full URL to allow this check: https://github.com/3scale/porta/pull/3103/files#diff-4f8bdb6882899231dfd4039aec865573cb6a79e61de460f4425e45b5aa966439R148

But in fact, I can adjust that check and maybe assume that if a path (and not the full URL) is used, then the request is for the OAS spec. Not sure if that makes sense though.
Then we wouldn't need to replace `/` with the `base_url`. And also we may skip setting the servers to `/` in the first place, because they are already like that in the JSON files (but I kind of prefered to make this even more explicit).

WDYT?

Reply:
@jlledom Yes, but as I explained previously, the full URL is needed for this check:  https://github.com/3scale/porta/pull/3103/files#diff-4f8bdb6882899231dfd4039aec865573cb6a79e61de460f4425e45b5aa966439R148

This is the way we know whether the request is for fetching the spec itself, or it's the API "try it out" call.

BTW, I recently saw that the `requestInterceptor` does distinguish calls for spec and API calls, I am wondering whether we could take advantage of that somehow, and pass this info from the requestInterceptor to the responseInterceptor.
But I am not sure I want to complicate the things even further...

Reply:
Hi, now I (think I) understand the problem. I tried rewriting the `autocompleteInterceptor` function like this:

```
export const autocompleteInterceptor = (response: SwaggerUIResponse, accountDataUrl: string, serviceEndpoint: string): Promise<Response> | SwaggerUIResponse => {
  try {
    const isSpec = !!JSON.parse(response.text).servers
    if (isSpec) {
      return response
    }
  // eslint-disable-next-line no-empty
  } catch {}

  return autocompleteOAS3(response, accountDataUrl, serviceEndpoint)
}
```

So the way to distinguish between spec and API responses is to deliberately look for something we expect to be in the spec response, for instance: it must be JSON and it must contain the servers attribute, but anything else could work as well. By doing it this way, I could:

- Remove the `update_base_url` method from `services_controller.rb`
- Simplify the `json` method from `services_controller.rb`
- Remove the specUrl parameter for `autocompleteInterceptor`

No sure if this is better 😅 , but at least `services_controller.rb` is simpler. I tried with two tenants and works fine. I didn't try the backend request bc I don't know how to fill the params.

>  But I am not sure I want to complicate the things even further...

Of curse you want.

Reply:
Well, I don't think this method is reliable @jlledom ...

The customer's API might have and endpoint returning a JSON, containing `"servers"` field, and then it would be a false positive for `isSpec`.